test/Mono.Linker.Tests.Cases/DataFlow/NullableAnnotations.cs

```diff
@@ -0,0 +1,289 @@
+		{
+			(Nullable.GetUnderlyingType (typeof (TNullable))).RequiresPublicProperties ();
+		}
+
+		[Kept]
+		static void UnderlyingTypeOfAnnotatedParameterRequiresPublicProperties ([KeptAttributeAttribute (typeof (DAM))][DAM (DAMT.PublicProperties)] Type tNullable)
+		{
+			(Nullable.GetUnderlyingType (tNullable)).RequiresPublicProperties ();
+		}
+
+		[Kept]
+		[ExpectedWarning ("IL2067")]
+		static void UnderlyingTypeOfUnannotatedParameterRequiresPublicProperties (Type tNullable)
+		{
+			(Nullable.GetUnderlyingType (tNullable)).RequiresPublicProperties ();
+		}
+
+		[Kept]
+		[ExpectedWarning ("IL2087")]
+		static void UnderlyingTypeOfUnannotatedGenericParameterRequiresProperties<TNullable> ()
+		{
+			(Nullable.GetUnderlyingType (typeof (TNullable))).RequiresPublicProperties ();
+		}
+
+		[Kept]
+		static void NullableOfAnnotatedGenericParameterRequiresPublicProperties<[KeptAttributeAttribute (typeof (DAM))][DAM (DAMT.PublicProperties)] T> () where T : struct
+		{
+			(Nullable.GetUnderlyingType (typeof (Nullable<T>))).RequiresPublicProperties ();
+		}
+
+		[Kept]
+		[ExpectedWarning ("IL2087")]
+		static void NullableOfUnannotatedGenericParameterRequiresPublicProperties<T> () where T : struct
+		{
+			(Nullable.GetUnderlyingType (typeof (Nullable<T>))).RequiresPublicProperties ();
+		}
+
+		[Kept]
+		static void NullableOfAnnotatedParameterRequiresPublicProperties ([KeptAttributeAttribute (typeof (DAM))][DAM (DAMT.PublicProperties)] Type t)
+		{
+			(Nullable.GetUnderlyingType (typeof (Nullable<>).MakeGenericType (t))).RequiresPublicProperties ();
+		}
+
+		[Kept]
+		static void TestGetUnderlyingTypeOnClasses ()
+		{
+			UnderlyingTypeOfAnnotatedParameterRequiresPublicProperties (typeof (TestClass));
+			UnderlyingTypeOfAnnotatedGenericParameterRequiresPublicProperties<TestClass> ();
+			UnderlyingTypeOfUnannotatedParameterRequiresPublicProperties (typeof (TestClass));
+			UnderlyingTypeOfUnannotatedGenericParameterRequiresProperties<TestClass> ();
+		}
+
+		[Kept]
+		static void TestGetUnderlyingTypeOnStructs ()
+		{
+			UnderlyingTypeOfAnnotatedParameterRequiresPublicProperties (typeof (TestStruct));
+			UnderlyingTypeOfAnnotatedGenericParameterRequiresPublicProperties<TestStruct> ();
+			UnderlyingTypeOfUnannotatedParameterRequiresPublicProperties (typeof (TestStruct));
+			UnderlyingTypeOfUnannotatedGenericParameterRequiresProperties<TestStruct> ();
+		}
+
+		[Kept]
+		static void TestGetUnderlyingTypeOnNullableStructs ()
+		{
+			UnderlyingTypeOfAnnotatedParameterRequiresPublicProperties (typeof (Nullable<TestStruct>));
+			UnderlyingTypeOfAnnotatedGenericParameterRequiresPublicProperties<Nullable<TestStruct>> ();
+			UnderlyingTypeOfUnannotatedParameterRequiresPublicProperties (typeof (Nullable<TestStruct>));
+			UnderlyingTypeOfUnannotatedGenericParameterRequiresProperties<Nullable<TestStruct>> ();
```

Comment:
While valid, the test is corrupted by the fact that we apply the annotation to `TestStruct` directly above. What we want to test is that passing `Nulable<TestStruct>` to annotated value applies the annotation on the `TestStruct` - but to do that we need to make sure that `TestStruct` is not marked by anything anywhere else in the test.

Reply:
You actually have tests like this below... I guess my only complaint is then the use of both TestStruct and TestClass - I don't see the need for that - they will do the exact same thing.